test/upgrade/fleet_test.go

```diff
@@ -0,0 +1,371 @@
+	if err != nil {
+		return err
+	}
+	if len(containers) == 0 {
+		return errors.New("no fleet container found")
+	}
+	port := containers[0].Ports[0].PublicPort
+	healthURL := fmt.Sprintf("http://localhost:%d/healthz", port)
+
+	retryStrategy := backoff.NewExponentialBackOff()
+	retryStrategy.MaxInterval = 1 * time.Second
+
+	if err := backoff.Retry(
+		func() error {
+			//nolint:gosec // G107: Ok to trust docker here
+			resp, err := http.Get(healthURL)
+			if err != nil {
+				return err
+			}
+			if resp.StatusCode != http.StatusOK {
+				return fmt.Errorf("non-200 status code: %d", resp.StatusCode)
+			}
+			return nil
+		},
+		retryStrategy,
+	); err != nil {
+		return fmt.Errorf("check health: %v", err)
+	}
+
+	return nil
+}
+
+func (f *Fleet) cleanup() {
+	output, err := f.execCompose(nil, "down", "-v", "--remove-orphans")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "stop fleet: %v %s", err, output)
+	}
+}
+
+func (f *Fleet) execCompose(env map[string]string, args ...string) (string, error) {
+
+	// docker compose variables via environment eg FLEET_VERSION_A
+	e := os.Environ()
+	for k, v := range env {
+		e = append(e, fmt.Sprintf("%s=%s", k, v))
+	}
+
+	// prepend default args
+	args = append([]string{
+		"compose",
+		"--project-name", f.ProjectName,
+		"--file", f.FilePath,
+	}, args...)
+
+	var stdout, stderr bytes.Buffer
+
+	cmd := exec.Command("docker", args...)
+	cmd.Env = e
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+	if err != nil {
+		return "", fmt.Errorf("docker: %v %s", err, stderr.String())
+	}
+
+	return stdout.String(), nil
+}
+
+// StartHost starts an osquery host using docker-compose and enrolls it with fleet.
+// Returns the container ID which is also the hostname and osquery host ID.
+func (f *Fleet) StartHost() (string, error) {
+	// get the enroll secret
+	client, err := f.Client()
+	if err != nil {
+		return "", err
+	}
+
+	enrollSecretSpec, err := client.GetEnrollSecretSpec()
+	if err != nil {
+		return "", err
+	}
+	if len(enrollSecretSpec.Secrets) == 0 {
+		return "", errors.New("no enroll secret found")
+	}
+
+	enrollSecret := enrollSecretSpec.Secrets[0].Secret
+
+	env := map[string]string{
+		"ENROLL_SECRET": enrollSecret,
+	}
+	output, err := f.execCompose(env, "run", "-d", "-T", "osquery")
+	if err != nil {
+		return "", err
+	}
+
+	// get the container id
+	containerID := output[:len(output)-1] // strip the newline from output
+
+	// inspect the container to get the hostname
+	containerJSON, err := f.dockerClient.ContainerInspect(context.Background(), containerID)
+	if err != nil {
+		return "", fmt.Errorf("inspect container: %v", err)
+	}
+	hostname := containerJSON.Config.Hostname
+
+	return hostname, nil
+}
+
+// Upgrade upgrades fleet to a specified version.
+func (f *Fleet) Upgrade(toVersion string) error {
+	env := map[string]string{
+		"FLEET_VERSION_B": toVersion + "asdf",
```

Comment:
Looks like the `asdf` was left behind from the demo :P

Reply:
yeah, accidentally committed